ambari-server/src/main/java/org/apache/ambari/server/state/cluster/ClusterImpl.java

```diff
@@ -1696,7 +1694,7 @@ public void deleteAllServices() throws AmbariException {
       }
 
       DeleteHostComponentStatusMetaData deleteMetaData = new DeleteHostComponentStatusMetaData();
-      for (Service service : services.values()) {
+      for (Service service : ImmutableList.copyOf(servicesById.values())) {
         deleteService(service, deleteMetaData);
```

Comment:
Why is defensive copy needed here? (just curiosity)

Reply:
Because modifies `servicesById` (and `services`, too).